icu4c/source/common/lstmbe.cpp

```diff
@@ -0,0 +1,792 @@
+                                                int32_t endPos,
+                                                UVector32 &foundBreaks ) const {
+    utext_setNativeIndex(text, startPos);
+    utext_moveIndex32(text, MIN_WORD_SPAN);
+    if (utext_getNativeIndex(text) >= endPos) {
+        return 0;       // Not enough characters for two words
+    }
+    utext_setNativeIndex(text, startPos);
+    UErrorCode status = U_ZERO_ERROR;
+
+    UVector32 offsets(status);
+    UVector32 indices(status);
+    fVectorizer->vectorize(text, startPos, endPos, offsets, indices, status);
+    int32_t* offsetsBuf = offsets.getBuffer();
+    int32_t* indicesBuf = indices.getBuffer();
+
+    int32_t input_seq_len = indices.size();
+    int32_t hunits = fData->forwardU().d1();
+
+    // To save the needed memory usage, the following is different from the
+    // Python or ICU4X implementation. We first perform the Backward LSTM
+    // and then merge the iteration of the forward LSTM and the output layer
+    // together because we only neetdto remember the h[t-1] for Forward LSTM.
+    Array1D c(hunits);
+
+    // TODO: limit size of hBackward. If input_seq_len is too big, we could
+    // run out of memory.
+    // Backward LSTM
+    Array2D hBackward(input_seq_len, hunits);
+    for (int32_t i = input_seq_len - 1; i >= 0; i--) {
+        Array1D hRow = hBackward.row(i);
+        if (i != input_seq_len - 1) {
+            hRow.assign(hBackward.row(i+1));
+        }
+        compute(fData->backwardW(), fData->backwardU(), fData->backwardB(),
+                fData->embedding().row(indicesBuf[i]),
+                hRow, c);
+    }
+
+    Array1D logp(4);
+
+    // Allocate fbRow and slice the internal array in two.
+    Array1D fbRow(2 * hunits);
+    Array1D forwardRow = fbRow.slice(0, hunits);  // point to first half of data in fbRow.
+    Array1D backwardRow = fbRow.slice(hunits, hunits);  // point to second half of data n fbRow.
+
+    // The following iteration merge the forward LSTM and the output layer
+    // together.
+    c.clear();  // reuse c since it is the same size.
+    for (int32_t i = 0; i < input_seq_len; i++) {
+        // Forward LSTM
+        // Calculate the result into forwardRow, which point to the data in the first half
+        // of fbRow.
+        compute(fData->forwardW(), fData->forwardU(), fData->forwardB(),
+                fData->embedding().row(indicesBuf[i]),
+                forwardRow, c);
+
+        // assign the data from hBackward.row(i) to second half of fbRowa.
+        backwardRow.assign(hBackward.row(i));
+
+        logp.dotProduct(fbRow, fData->outputW()).add(fData->outputB());
+
+        // current = argmax(logp)
+        LSTMClass current = (LSTMClass)logp.maxIndex();
+        // BIES logic.
+        if (current == BEGIN || current == SINGLE) {
+            if (i != 0) {
+                foundBreaks.addElement(offsetsBuf[i], status);
+            }
+        }
```

Comment:
Thought: This is where we could include a BIES normalizer algorithm, because this could result in a break in the middle of a grapheme cluster if code point mode is used (this is fairly rare but can happen, especially on large texts).

Reply:
currently, the manual_predict, I try to mimic from the python code does not call the bies normalization and I intend to keep a bug-to-bug compatibility during the porting.